providers/base/bin/cpuid.py

```diff
@@ -189,23 +372,41 @@ def __del__(self):
 
 
 def main():
+    parser = argparse.ArgumentParser(
+                    prog='cpuid',
+                    description='Identifies the Intel generation based on cpu \
+                                ids')
+    parser.add_argument('--intel_gen_number', action='store_true')
+    args = parser.parse_args()
+
     cpuid = CPUID()
     cpu = cpuid(1)
 
     # Lets play Guess The CPU!
     # First lets get the name from /proc/cpuinfo
     cpu_data = check_output('lscpu', universal_newlines=True).split('\n')
     for line in cpu_data:
-        if line.startswith('Model name:'):
+        if line.startswith('Model name:') and not args.intel_gen_number:
             print("CPU Model: %s" % line.split(':')[1].lstrip())
 
     my_id = (hex(cpu[0]))
     complete = False
-    for key in CPUIDS.keys():
-        for value in CPUIDS[key]:
-            if value in my_id:
-                print("CPUID: %s which appears to be a %s processor" %
-                      (my_id, key))
+    for key in CPUDICT.keys():
+        platform = CPUDICT[key]
+        platform_ids = platform['cpuids']
+        for platform_id in platform_ids:
+            if platform_id.lower() in my_id:
+                if args.intel_gen_number:
+                    if platform['chiptype'] == ChipType.CLIENT and \
+                       platform['generation'] > 0:
+                        print(platform['generation'])
+                    else:
+                        print("%s is not supported for conversion to Intel \
+                              generation number")
+                else:
+                    print("CPUID: %s which appears to be a %s processor" %
+                          (my_id, key))
+
```

Comment:
With this complexity this whole function is not just a simple entry point to the program. There's a lot of complex logic that is implemented here and that logic needs to be outlined (possibly to more than one function). The aforementioned decomposition will also facilitate unit testing the new branches of logic that are proposed here which are missing and necessary.